src/proc/mod.rs

```diff
@@ -212,7 +213,131 @@ impl crate::Binding {
+    /// binding that can vary, and everything else either defaults to flat, or
+    /// requires an explicit flat qualifier/attribute/what-have-you.
+    ///
+    /// [`Binding`s]: super::Binding
+
+    pub fn apply_common_default_interpolation(&mut self) {
+        use super::{Binding, ScalarKind, Type, TypeInner};
+
+        fn type_is_variable(ty: Handle<Type>, arena: &Arena<Type>) -> bool {
+            match &arena[ty].inner {
+                &TypeInner::Scalar { kind: ScalarKind::Float, width: 4 } |
+                &TypeInner::Vector { kind: ScalarKind::Float, width: 4, .. } => true,
+                &TypeInner::Struct { ref members, .. } =>
+                    members.iter().all(|m| type_is_variable(m.ty, arena)),
+                _ => false,
+            }
+        }
+
+        fn default_binding(binding: &mut Binding, ty: Handle<Type>, types: &Arena<Type>) {
+            if let Binding::Location { ref mut interpolation, ref mut sampling, .. } = *binding {
+                if interpolation.is_none() {
+                    if type_is_variable(ty, types) {
+                        *interpolation = Some(super::Interpolation::Perspective);
+                    } else {
+                        *interpolation = Some(super::Interpolation::Flat);
+                        *sampling = None;
+                    }
+                }
+
+                if sampling.is_none() && *interpolation != Some(super::Interpolation::Flat) {
+                    *sampling = Some(super::Sampling::Center);
+                }
+            }
+        }
+
+        /// Choose an appropriate default interpolation for `binding`, or if
+        /// that is absent, for the `Binding` of the members of `ty`, which must
+        /// be a struct.
+        ///
+        /// Since we must consult unknown parts of `types` just to choose an
+        /// appropriate default, Rust won't let us mutate a `Binding` that
+        /// belongs to one of its members, so we also contribute to a list of
+        /// deferred mutations.
+        fn default_binding_or_struct(binding: &mut Option<Binding>,
+                                     ty: Handle<Type>,
+                                     types: &Arena<Type>,
+                                     deferred: &mut Vec<(Handle<Type>, Vec<Binding>)>)
+        {
+            if let Some(binding) = binding.as_mut() {
+                default_binding(binding, ty, types);
+            } else if let TypeInner::Struct { ref members, .. } = types[ty].inner {
+                let new_ty_bindings = members.iter()
+                    .map(|member| {
+                        // unwrap: all members of structs used as `EntryPoint`
+                        // arguments or return values must have bindings.
+                        let mut binding_copy = member.binding.as_ref().unwrap().clone();
+                        default_binding(&mut binding_copy, member.ty, types);
```

Comment:
Do we actually need to resolve it here? That forces us to carry around `Vec<Binding>`. Instead, since we are iterating the members anyway (when going through the deferred list), we could then call `default_binding` in place.

So we really need just a set of Type handles relevant to I/O structs here. It could be a `BitSet`, where bit N corresponds to `type_handle.index() == N`.

Reply:
The deferred set will always need to somehow record how each member is to be defaulted. If we have mutable access to the `StructMember`, then by definition we don't have any access to its type at all, but we need to check the type to decide how to default it.

Reply:
Some members need to be defaulted to perspective interpolation, others to flat, based on their type. So the result of the scan needs to be recording information about individual members; a `BitSet` of type indices isn't enough.

Reply:
> If we have mutable access to the StructMember, then by definition we don't have any access to its type at all

I don't understand this part. If you have `member: StructMember`, then `module.types[member.ty]` would give you a type that you can check to figure out the default interpolation

Reply:
My apologies, I understand it now. We can come back to this later.
It's a bit unfortunate that this processor would have a vec of a vec of things, but it's not a big concern, really.

Reply:
Actually, I found a nice solution for this. No allocations at all, just some surprising mutations.